### PR TITLE
Ports Goonchat

### PR DESCRIPTION
**Pull request by Cyberboss**
Follow up to #26542 

See the latest commit, logs all requests to and from the script. Goal is to get this testmerged and have clients use the `chat_broke` verb when it breaks so we can narrow down the cause

### CODE REVIEW

**Comment by WJohn** [comment]
You know, the last one never really stated what it changes.

Is it just a refactor? Does it bring any new features players will notice?

**Comment by IlyaIU** [comment]
@WJohn 
>It's a HTML-based chat. It supports UTF-8, custom fonts, custom RMB actions (adminpanel on name RMB click, anyone?) and all kinds of fancy stuff.

**Comment by IlyaIU** [comment]
image
image

**Comment by IlyaIU** [comment]
image
and what Paradise done to Goonchat

**Comment by ChangelingRain** [comment]
Service's chat color isn't working, it's just black.

**Comment by ChangelingRain** [comment]
Supply too.

**Comment by ChangelingRain** [comment]
I'm not getting some messages and I'm not sure why. Stuff like whispers from not-nearby mobs as a ghost, adminhelps(but the ticket message shows up...)
It's odd, probably missed <<s

**Comment by 4dplanner** [comment]
Note sure if this has been noted, but the link to this PR under "the following PRs are testmerged" isn't working - presumably as a result of this PR

Links to http://%3Ca%20href%3Dhttps//github.com/tgstation/tgstation/pull/26965%3Ehttps://github.com/tgstation/tgstation/pull/26965%3C/a%3E

**Comment by 4dplanner** [comment]
Though every other link is working, so it might actually just be the recent test merge changes

File: code/modules/goonchat/browserassets/js/browserOutput.js [hunk 1]
```diff
@@ -0,0 +1,910 @@
...
+			$('#ping').slideDown('fast');
+			opts.pingDisabled = false;
+		} else {
+			$('#ping').slideUp('fast');
+			opts.pingDisabled = true;
+		}
+		setCookie('pingdisabled', (opts.pingDisabled ? 'true' : 'false'), 365);
+	});
+
+	$('#saveLog').click(function(e) {
```
Comment (SpaiR) [inline code/modules/goonchat/browserassets/js/browserOutput.js]: @Cyberboss This shit won't work in any case. If you try `Save log` button sync ajax won't happen, but the client will experience lag, which will require restarting of the game. Async ajax would help, but in that case there is need of callback, to use css file. This bug has almost every build with goonchat, even goon themselves. And yes, problem doesn't connected with `window.XMLHtpRequest` type error.  So, you may be intrested in this decision: https://github.com/TauCetiStation/TauCetiClassic/pull/1424/commits/27d18e8e203f325a982196e8490b0f9edbdf112b
It works as well as it should. Also it fixed amazing `win.document.body.innerHTML = saved;`, where `saved` var has `<style>...</style>`, which won't work in `body` tag. Furthermore, there you can see better variant to center chat log window on the screen.

File: code/modules/mob/living/silicon/robot/robot_defense.dm [hunk 2]
```diff
@@ -152,7 +152,7 @@
 	sleep(5)
 	to_chat(src, "<span class='danger'>Would you like to send a report to NanoTraSoft? Y/N</span>")
 	sleep(10)
-	to_chat(src, "<span class='danger'>> N</span>")
```
Comment (ChangelingRain) [inline code/modules/mob/living/silicon/robot/robot_defense.dm]: This was not an error

Comment (Cyberboss) [inline code/modules/mob/living/silicon/robot/robot_defense.dm]: whoops

**Comment by ChangelingRain** [comment]
Is this ready yet

**Comment by Cyberboss** [comment]
I gotta fix the radio channels, remove the debug code and run another tm

**Comment by Cyberboss** [comment]
Gonna finish TGS3 before I resume work on this, if someone wants to take it from me go ahead

**Comment by optimumtact** [comment]
Test merge

**Comment by Cyberboss** [comment]
No, the bug showed up again. I think something about the debug code I had
in is fixed it.

**Comment by optimumtact** [comment]
well

**Comment by Cyberboss** [comment]
/me slinks back into the shadows for serveral months

**Comment by Tacolizard** [comment]
is it fixed?

**Comment by Cyberboss** [comment]
It's cursed

**Comment by Tacolizard** [comment]
what's the merge conflict?

File: code/modules/goonchat/browserOutput.dm [hunk 3]
```diff
@@ -215,7 +215,7 @@ GLOBAL_DATUM_INIT(iconCache, /savefile, new("data/iconCache.sav")) //Cache of ic
 	
 	if(islist(target))
 		for(var/I in target)
-			to_chat(target, message)
```
Comment (optimumtact) [inline code/modules/goonchat/browserOutput.dm]: nice

Comment (Tacolizard) [inline code/modules/goonchat/browserOutput.dm]: kek

**Comment by Cyberboss** [comment]
Todo:

- Some clock shit isn't big for some reason

- Fix external hrefs

**Comment by optimumtact** [comment]
So what was the bug

**Comment by Cyberboss** [comment]
no fucking clue

**Comment by optimumtact** [comment]
:thinking:

**Comment by thefastfoodguy** [comment]
SPEEDMERGE THIS!

**Comment by lzimann** [comment]
So, what COULD possibily be bug:
Using `usr << browse(dat, "options")` without setting `window=name` in the options is going to override the chat's output(the default, iirc, element on skin.dmf). 

I am still not sure how it would even make everyone randomly fail to load the chat or fail to set it.

**Comment by Tacolizard** [comment]
Misc bugs:

If you `Say` HTML tags, the opening tag will be spoken but the closing tag is removed

Spaces are prepended and appended to every message

Using the high volume function on a headset only bolds the text, it doesn't enlarge it

**Comment by Cyberboss** [comment]
The main issue with small text was the stylesheet being incorrect, should all be fixed now

**Review by silicons — COMMENTED** [review]
:3

**Comment by Tacolizard** [comment]
I don't know how hard/easy it would be, but it would probably be a good idea to remove custom fonts, because we use fonts to denote what language someone is speaking (silicons use a monospace font, but you can use goonchat to set everything to monospace).

**Comment by Cyberboss** [comment]
Already done

**Comment by Tacolizard** [comment]
Oh, sorry I didn't see the commit

**Comment by Iamgoofball** [comment]
@Cyberboss 
You removed the custom font support? Why?

File: code/modules/goonchat/browserassets/html/browserOutput.html [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+		</div>
+		<div id="options">
+			<a href="#" class="toggle" id="toggleOptions" title="Options"><i class="icon-cog"></i></a>
+			<div class="sub" id="subOptions">
+				<a href="#" class="decreaseFont" id="decreaseFont"><span>Decrease font size</span> <i class="icon-font">-</i></a>
+				<a href="#" class="increaseFont" id="increaseFont"><span>Increase font size</span> <i class="icon-font">+</i></a>
+				<a href="#" class="togglePing" id="togglePing"><span>Toggle ping display</span> <i class="icon-circle"></i></a>
+				<a href="#" class="highlightTerm" id="highlightTerm"><span>Highlight string</span> <i class="icon-tag"></i></a>
+				<a href="#" class="saveLog" id="saveLog"><span>Save chat log</span> <i class="icon-save"></i></a>
+				<a href="#" class="clearMessages" id="clearMessages"><span>Clear all messages</span> <i class="icon-eraser"></i></a>
```
Comment (Iamgoofball) [inline code/modules/goonchat/browserassets/html/browserOutput.html]: Literally why did you remove the custom font option, I demand a reason and why I shouldn't re-add it.

We don't differentiate between languages with fonts anymore, either. We use icons for it.

File: code/modules/goonchat/browserassets/js/browserOutput.js [hunk 5]
```diff
@@ -0,0 +1,904 @@
...
+
+	$('#increaseFont').click(function(e) {
+		var fontSize = parseInt($messages.css('font-size'));
+		fontSize = fontSize + 1 + 'px';
+		$messages.css({'font-size': fontSize});
+		setCookie('fontsize', fontSize, 365);
+		internalOutput('<span class="internal boldnshit">Font size set to '+fontSize+'</span>', 'internal');
+	});
+
+	$('#chooseFont').click(function(e) {
```
Comment (Iamgoofball) [inline code/modules/goonchat/browserassets/js/browserOutput.js]: Re-enable this.

**Comment by Tacolizard** [comment]
@Iamgoofball 

>I don't know how hard/easy it would be, but it would probably be a good idea to remove custom fonts, because we use fonts to denote what language someone is speaking (silicons use a monospace font, but you can use goonchat to set everything to monospace).

**Comment by Iamgoofball** [comment]
You're incredibly wrong 

I added that shit

**Comment by Tacolizard** [comment]
@Iamgoofball You added what?

File: code/game/gamemodes/gang/gang_datum.dm [hunk 6]
```diff
@@ -246,47 +246,90 @@
...
+		pay_territory_income_to_soldiers(sbonus)
+		pay_all_clothing_bonuses()
+		announce_all_influence()
+
+/datum/gang/proc/pay_all_clothing_bonuses()
+	for(var/datum/mind/mind in gangsters|bosses)
+		pay_clothing_bonus(mind)
+
+/datum/gang/proc/pay_clothing_bonus(var/datum/mind/gangsta)
+	var/mob/living/carbon/human/gangbanger = gangsta.current
```
Comment (Iamgoofball) [inline code/game/gamemodes/gang/gang_datum.dm]: remember to put dis at da top of da list

**Comment by Iamgoofball** [comment]
@Cyberboss still actively ignoring me

**Comment by ChangelingRain** [comment]
Well, if the bug is fixed, then I guess it's time to do the mergeful.

**Comment by lzimann** [comment]
It's not even sure that it is fixed. It is a possible cause but most likely not.

**Comment by PJB3005** [comment]
> didn't port libvg.
